feat: Support EventMetadata filter for EventsQuery/TrendsQuery

danielbachhuber opened the pull request:
See https://github.com/PostHog/posthog/issues/29881
See https://posthog.slack.com/archives/C0368RPHLQH/p1742417106336409

## Changes

Adds support for filtering `EventsQuery` or `TrendsQuery` by a new `EventMetadata` filter:

https://github.com/user-attachments/assets/31289ade-01c4-4395-ad95-3ef786259f82

The happy path works but I'm not sure how many gotchas I might be missing.
## How did you test this code?

Added backend tests.

danielbachhuber committed:
First pass at event metadata filter support

danielbachhuber committed:
Dynamically add group types

danielbachhuber committed:
Display the icon

danielbachhuber committed:
Fix type issue

danielbachhuber committed:
Fix other type issue

mariusandra commented on frontend/src/lib/components/TaxonomicFilter/taxonomicFilterLogic.tsx:
```diff
@@ -593,6 +605,24 @@ export const taxonomicFilterLogic = kea<taxonomicFilterLogicType>([
+            (s) => [s.groupTypes],
+            (groupTypes) => {
+                const options = [
+                    { icon: <IconLogomark />, name: 'Distinct ID', value: 'distinct_id' },
+                    { icon: <IconLogomark />, name: 'Timestamp', value: 'timestamp' },
```
Can you also add `event` here? Perhaps also `person_id` if it works well.

mariusandra replied:
Thinking out loud...

While I think it makes total sense to have a new event metadata `TaxonomicFilterGroupType` for this new tab/pill... I wonder if this is the right `PropertyFilterType` to add. It feels a bit restrictive. More specifically, I wonder if `EventMetadataPropertyFilter` could be merged with `HogQLPropertyFilter`.

Looking at this list of options here, we effectively have `expression operator value`, where the expression comes from a predefined list of SQL expressions. There's no backend validation... so it's effectively a generic `SQLValueFilter` or something like that.

The existing `HogQLPropertyFilter` does the `expression` part as its `key`. We could just add `operator` and `value` into it... or if making a new property filter, to make it more abstract in its name.

danielbachhuber replied:
> Can you also add `event` here? Perhaps also `person_id` if it works well.

Added with 94d3026689. Both work fine.

> The existing `HogQLPropertyFilter` does the `expression` part as its `key`. We could just add `operator` and `value` into it... or if making a new property filter, to make it more abstract in its name.

Doesn't that mean we'd need to apply [this logic](https://github.com/PostHog/posthog/blob/94d3026689560c309bea5b35cdba87a836cd64df/posthog/hogql/property.py#L446-L467) and [this logic](https://github.com/PostHog/posthog/blob/94d3026689560c309bea5b35cdba87a836cd64df/posthog/hogql/property.py#L160-L254) in the client?

mariusandra replied:
> Doesn't that mean we'd need to apply [this logic](https://github.com/PostHog/posthog/blob/94d3026689560c309bea5b35cdba87a836cd64df/posthog/hogql/property.py#L446-L467) and [this logic](https://github.com/PostHog/posthog/blob/94d3026689560c309bea5b35cdba87a836cd64df/posthog/hogql/property.py#L160-L254) in the client?

Hmm... well, no... it'd work exactly like the filter you now added works, just renamed to `HogQLPropertyValueFilter` or something more abstract.

mariusandra commented on posthog/hogql/property.py:
```diff
@@ -549,6 +551,48 @@ def property_to_expr(
             ),
             right=ast.Constant(value=cohort.pk),
         )
+    elif property.type == "event_metadata":
+        if scope != "event":
+            raise NotImplementedError(f"The '{property.type}' property filter does not work in '{scope}' scope")
+        operator = cast(Optional[PropertyOperator], property.operator) or PropertyOperator.EXACT
+        value = property.value
+
+        if isinstance(value, list):
+            if len(value) == 0:
+                return ast.Constant(value=1)
+            elif len(value) == 1:
+                value = value[0]
+            else:
+                # Using an AND here instead of `in()` or `notIn()`, due to Clickhouses poor handling of `null` values
+                exprs = [
+                    property_to_expr(
+                        Property(
+                            type=property.type,
+                            key=property.key,
+                            operator=property.operator,
+                            value=v,
+                        ),
+                        team,
+                        scope,
+                    )
+                    for v in value
+                ]
+                if (
+                    operator == PropertyOperator.NOT_ICONTAINS
+                    or operator == PropertyOperator.NOT_REGEX
+                    or operator == PropertyOperator.IS_NOT
+                ):
+                    return ast.And(exprs=exprs)
+                return ast.Or(exprs=exprs)
```
This and/or looks like a duplicate of the above block with
```
        property.type == "event"
        or property.type == "feature"
        or property.type == "person"
        or property.type == "group"
        or property.type == "data_warehouse"
        or property.type == "data_warehouse_person_property"
        or property.type == "session"
        or property.type == "recording"
        or property.type == "log_entry"`
```

Could this logic just be merged there instead?

danielbachhuber replied:
> Could this logic just be merged there instead?

Yeah, good catch. Updated with b68ac92ea9

mariusandra reviewed (commented):
Cool stuff, left some thoughts inline.

danielbachhuber committed:
Consolidate logic

danielbachhuber committed:
Add `event` and `person_id`

danielbachhuber committed:
Hacky workaround for type issue

aspicer commented:
Do we want to call this "$group_2" in the customer facing UI? I don't think we use that terminology anywhere and it's quite an implementation detail to expose.

"Organization ID" might make more sense? I'm not sure we want to surface that group naming anywhere?

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

aspicer commented:
This looks awesome. Much overdue.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

danielbachhuber commented:
> Do we want to call this "$group_2" in the customer facing UI? I don't think we use that terminology anywhere and it's quite an implementation detail to expose.
> 
> "Organization ID" might make more sense? I'm not sure we want to surface that group naming anywhere?

@aspicer My intent was to ease this transition:
![CleanShot 2025-03-20 at 16 38 47](https://github.com/user-attachments/assets/08fe7f40-9b0f-42f3-900f-5d173ac50b7e)

Specifically, when the user selects "organization" in the UI and then sees `$group_2` in the filter.

It looks like `getCoreFilterDefinition()` is where the label is fetched:

https://github.com/PostHog/posthog/blob/55db33c8dee2714cc24a0bc28632c393d92cb3fc/frontend/src/lib/components/PropertyKeyInfo.tsx#L37
https://github.com/PostHog/posthog/blob/55db33c8dee2714cc24a0bc28632c393d92cb3fc/frontend/src/lib/taxonomy.tsx#L1915-L1921

We could fetch the group type mapping labels dynamically but I don't love the idea of adding a call to a logic. It seems like too many of our components have random dependencies on logics.

I do think `$group_2` is the value we should store in the filter, though (not `group_type`).

Thoughts?

mariusandra commented:
It's definitely better for the user to see a clearly labelled "Organization ID" rather than a cryptic "$group_2", so I think it's worth taking whatever extra steps are needed to make it work. We should have all the group names in memory normally as well.

mariusandra reviewed (changes requested):
I had a play around. While it did filter when I entered stuff to the fields, there's no autocomplete:

<img width="995" alt="image" src="https://github.com/user-attachments/assets/8733451d-4d08-44b2-8d76-5cd0d965efb5" />

Plus the timestamp field is not treated as a datetime field (no date/time controls, before/after, etc)

<img width="954" alt="image" src="https://github.com/user-attachments/assets/ac0847e5-a11f-494d-a822-15cb5dc963f6" />


This is what I'd expect:

<img width="721" alt="image" src="https://github.com/user-attachments/assets/ecaa202d-df83-48d6-ae51-fe2dbbda4f34" />

<img width="751" alt="image" src="https://github.com/user-attachments/assets/e798ff8b-9222-4ea0-aa07-f3ddf0217597" />

danielbachhuber committed:
Merge branch 'master' into crm/event-metadata

danielbachhuber committed:
Reorder

danielbachhuber committed:
Persist the group type name as the label

danielbachhuber committed:
Display definition popover

danielbachhuber committed:
Monkey-patch eventMetadata into property definitions

danielbachhuber committed:
Special case behavior for the group type name

danielbachhuber committed:
Connect auto-suggest for properties

danielbachhuber commented:
_mumble mumble last 20%_ 😅

> While it did filter when I entered stuff to the fields, there's no autocomplete:

Connected auto-complete with cab879c96a101acfd54fb6a5e19684364ba316d2

> Plus the timestamp field is not treated as a datetime field (no date/time controls, before/after, etc)

I had to macgyver eventMetadata into property definitions for this one: e5f6405b7a30d01af9605d3cb7be68e54513ee9f

Albeit a little bit funky, I think it's better than: 1) divergent code paths in the components, and 2) creating an API endpoint with a static response.

Bonus is that the popovers look nice now:

![CleanShot 2025-03-21 at 10 32 36@2x](https://github.com/user-attachments/assets/a71411a9-9adc-4c2e-b380-046d61ccdf60)

danielbachhuber committed:
Fix type issues

danielbachhuber committed:
Merge branch 'master' into crm/event-metadata

danielbachhuber committed:
Merge branch 'master' into crm/event-metadata

danielbachhuber committed:
Fix LLM hallucination

danielbachhuber committed:
Ensure eventMetadata property definitions are available early

danielbachhuber committed:
Don't load for date picker

mariusandra commented on frontend/src/lib/components/PropertyFilters/components/TaxonomicPropertyFilter.tsx:
```diff
@@ -144,6 +144,7 @@ export function TaxonomicPropertyFilter({
                         value: newValue || null,
                         operator: newOperator,
                         type: filter?.type,
+                        label: filter?.label,
```
The trouble with the `label` approach is that once someone changes the group's key (e.g. "organization" -> "company"), all old filters will be invalid. That's why we fetch the group/action/cohort/etc name in most places.

danielbachhuber replied:
Yes, I'm aware. I decided the benefit significantly outweighed the cost.

mariusandra commented on frontend/src/models/propertyDefinitionsModel.ts:
```diff
@@ -107,18 +108,32 @@ const checkOrLoadPropertyDefinition = (
-                    ?.actions.loadPropertyDefinitions([propertyName], definitionType, groupTypeIndex),
-            0
-        )
+        if (definitionType === PropertyDefinitionType.EventMetadata) {
+            propertyDefinitionsModel
+                .findMounted()
+                ?.actions.updatePropertyDefinitions(
+                    Object.fromEntries(
+                        eventMetadataPropertyDefinitions.map((definition) => [
+                            `${PropertyDefinitionType.EventMetadata}/${definition.id}`,
+                            definition,
+                        ])
+                    )
+                )
```
It's loading all sorts of random stuff now.

![image](https://github.com/user-attachments/assets/8f88c2bc-3d00-4bc2-af04-37d9ba5ff367)

mariusandra reviewed (changes requested):
Thanks for all the work so far! 

I had a play around and would still like some changes/fixes. The biggest one is that the autocomplete still doesn't work. I see that the backend is there, and looking at the tests it seems to do exactly what's needed... but the code just branches off in different directions. First the parameter isn't a boolean (so passing in `is_column=false` will always fail), 

![image](https://github.com/user-attachments/assets/a62f9446-b560-4f39-a07a-d628d1658c7b)

but then it just doesn't work and keeps adding "properties":

![image](https://github.com/user-attachments/assets/97f2972a-1f30-4960-a1ac-9e70e0b5526c)

...

Then, I can sort of get why you'd link this to the `propertyDefinitionsModel` (to easily get groups, plus the type), but I would seriously try uncoupling them. All our "models" effectively reflect some database objects. This feels like a roundabout way of adding static data to a list. I'd rather have a list within the taxonomic filter logic, which already has access to groups.

Plus there are side effects now, e.g. every time I open "event", we try to load the definition for the property "event" and many others, making a lot of log noise and requests:

![image](https://github.com/user-attachments/assets/f336f4e4-1a9b-4086-8385-a00b1888b3ce)

If the goal is to get the type (datetime?) and the definition popup to show data, there's hopefully a cleaner way. Going through `checkOrLoadPropertyDefinition` for static data seems excessive 🫠

danielbachhuber commented:
> The biggest one is that the autocomplete still doesn't work. I see that the backend is there, and looking at the tests it seems to do exactly what's needed... but the code just branches off in different directions. First the parameter isn't a boolean (so passing in `is_column=false` will always fail),

@mariusandra Are you sure the logic was regenerated when you checked out the branch? It works fine for me:

![CleanShot 2025-03-24 at 04 13 36@2x](https://github.com/user-attachments/assets/2935c85c-6d9d-4f37-bc63-5ef2f23485ff)

> Plus there are side effects now, e.g. every time I open "event", we try to load the definition for the property "event" and many others, making a lot of log noise and requests:

I'm not seeing that...

![CleanShot 2025-03-24 at 04 14 48@2x](https://github.com/user-attachments/assets/2bc920c0-4dbe-4759-b0ea-813abb7b7c05)

> Then, I can sort of get why you'd link this to the `propertyDefinitionsModel` (to easily get groups, plus the type), but I would seriously try uncoupling them. All our "models" effectively reflect some database objects. This feels like a roundabout way of adding static data to a list. I'd rather have a list within the taxonomic filter logic, which already has access to groups.

It's actually quite a bit more. Most of the components are tightly coupled to property definitions. Refactoring away would mean some serious reinventing of the wheel. I think the one change I would be open to making would be to move the event metadata property definitions to a server endpoint so I [can avoid this weird branch](https://github.com/PostHog/posthog/blob/96770d6402fa3156c59dfcec9f54768d9431b5e4/frontend/src/models/propertyDefinitionsModel.ts#L117-L136).

> If the goal is to get the type (datetime?) and the definition popup to show data, there's hopefully a cleaner way. Going through `checkOrLoadPropertyDefinition` for static data seems excessive 🫠

But the problem is that it's not "static data". It's a little bit of static data (the events table columns, whose types will never change), and then some dynamic data.

danielbachhuber committed:
Merge branch 'master' into crm/event-metadata

mariusandra commented:
Very weird, everything indicates that I am on the latest branch. 

![image](https://github.com/user-attachments/assets/8cac3497-0350-4697-9249-24a8ca9cfdfd)

![image](https://github.com/user-attachments/assets/a7b2f15f-d619-4ad5-9642-d5435215630c)

![image](https://github.com/user-attachments/assets/fc547682-fc3a-4184-ba6e-f61985813dd2)

However it doesn't work:

![2025-03-24 12 38 43](https://github.com/user-attachments/assets/02e9aaf8-f9d9-476a-a772-60aca30b2abf)

cause the backend runs [this code](https://github.com/PostHog/posthog/blob/96770d6402fa3156c59dfcec9f54768d9431b5e4/posthog/api/event.py#L348-L356):

![image](https://github.com/user-attachments/assets/ab5c5e93-3e9a-453b-a99c-78d4ad1a8112)

Are you sure every change is pushed? 🤔

mariusandra commented:
I think I see what's happening: I hadn't run the plugin server and had events > 7 days old, so it didn't find any values to return... but then I also ran into a bug with the "value" parameter when I tried searching for something...

danielbachhuber commented:
> but then I also ran into a bug with the "value" parameter when I tried searching for something

@mariusandra Yeah, I see what this bug is. Fixing...

danielbachhuber committed:
Properly handle searching

mariusandra commented:
Regarding the extra queries to fetch properties, it seems that it makes one when I load the page:

![2025-03-24 12 52 43](https://github.com/user-attachments/assets/a63da712-2e4e-46ad-a1ff-125c7b9c3edf)

and I'm not too sure what's adding all these calls here, but something's definitely trying to load it as a property if I look at the kea actions log

![2025-03-24 12 55 39](https://github.com/user-attachments/assets/b98c701c-1231-4899-bb1e-f31ea3f9eeba)

danielbachhuber committed:
Fix missing argument

danielbachhuber commented:
@mariusandra Searching fixed with bdb88ea8558523b7484677b8db8210110f7ca01d

The property definition issue you're seeing might've been 21576b309865b4cde255d8673059234fdea49569. I didn't see it, but the type system caught it and it was definitely a bug. Can you retest?

danielbachhuber committed:
Fix hallucination

Co-authored-by: Marius Andra <marius.andra@gmail.com>

danielbachhuber commented:
> Regarding the extra queries to fetch properties, it seems that it makes one when I load the page:

@mariusandra I think I see this now when selecting a filter. Debugging...

---

> Then, I can sort of get why you'd link this to the propertyDefinitionsModel (to easily get groups, plus the type), but I would seriously try uncoupling them. All our "models" effectively reflect some database objects. This feels like a roundabout way of adding static data to a list. I'd rather have a list within the taxonomic filter logic, which already has access to groups.

Going back to this point, how strongly do you feel about this?

danielbachhuber committed:
Ensure type is passed through to avoid default 'event' type

danielbachhuber committed:
Ensure `propertyDefinitionType` is passed through in more places

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

danielbachhuber committed:
Just a 'Type'

danielbachhuber committed:
Handle `is_column=false`

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

danielbachhuber commented:
> I'm willing to 🙈 on the part of it all going through property definitions, however I'd see if there's any way to avoid those "loading"/"missing" states (those kea actions in one of the last gifs).

@mariusandra I think I fixed this with fcfba126d322d171f4ed167b3a02caa14dd74ed7 and 0fda1bc35814a5290a3005cca1e7f8503b64fdcb (at least, I don't see the erroneous `property_definitions?limit=50&properties=event&type=event` API request anymore).

> Perhaps avoid going through checkOrLoadPropertyDefinition altogether?

I tried this initially with e5f6405b7a30d01af9605d3cb7be68e54513ee9f but had to change it in bef7375bfb9d95aa11fb0558be2e5f5ab7140e77 because of some bug (don't recall exactly). Is there some alternative you had in mind?

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

danielbachhuber commented on frontend/src/lib/components/DefinitionPopover/DefinitionPopoverContents.tsx:
```diff
@@ -337,6 +337,24 @@ function DefinitionView({ group }: { group: TaxonomicFilterGroup }): JSX.Element
+            <>
+                {sharedComponents}
+                <DefinitionPopover.Grid cols={2}>
+                    <DefinitionPopover.Card title="Property Type" value={_definition.property_type ?? '-'} />
```
df418a0309ee1ebc690110019086bf1e1b46e7b9

mariusandra commented on frontend/src/lib/components/TaxonomicFilter/taxonomicFilterLogic.tsx:
```diff
@@ -319,6 +327,25 @@ export const taxonomicFilterLogic = kea<taxonomicFilterLogicType>([
+                            return coreDefinition ? coreDefinition.label : option.name
+                        },
+                        getValue: (option: PropertyDefinition) => option.id,
+                        valuesEndpoint: (key) => {
+                            return `api/event/values/?key=${key}&is_column=true`
+                        },
```
It might be wise to sanitize the parameter here

danielbachhuber replied:
e09495dc0ee35c789d65c95226f568fcecdc61d2

mariusandra commented on posthog/api/event.py:
```diff
@@ -306,12 +307,15 @@ def values(self, request: request.Request, **kwargs) -> response.Response:
-                ast.CompareOperation(
-                    op=ast.CompareOperationOp.NotEq,
-                    left=ast.Field(chain=["properties", key]),
-                    right=ast.Constant(value=None),
-                ),
             ]
+            if not is_column:
+                conditions.append(
+                    ast.CompareOperation(
+                        op=ast.CompareOperationOp.NotEq,
+                        left=ast.Field(chain=["properties", key]),
+                        right=ast.Constant(value=None),
+                    )
+                )
```
It's probably good for consistency to use the `chain` variable from below here as well.

danielbachhuber replied:
Hm. The left side won't ever be `[key]`, though. I think it's better to have it as-is?

mariusandra replied:
It may be worth keeping the non-null check as a default like it used to be (so removing the `if`)... but maybe not. I'm quite indifferent actually, so up to you

danielbachhuber replied:
I'm indifferent, too. Added it with 9bf1d91a0d

mariusandra reviewed (changes requested):
The person_id ~filter~ dropdown doesn't work:

![image](https://github.com/user-attachments/assets/5b9a3e83-8dd5-4b17-b9de-ca5ef1cb5705)

and I left a few other comments inline. 

As for skipping `checkOrLoadPropertyDefinition`, well, just prepopulate `propertyDefinitionStorage` with the metadata properties and that should be enough, no? One trick you can use is to rename the `propertyDefinitionStorage` reducer to `rawPropertyDefinitionStorage`, and then add a new selector called `propertyDefinitionStorage` that merges `rawPropertyDefinitionStorage` and `eventMetadataPropertyDefinitions` into one object.

danielbachhuber committed:
Handle `person_id` auto-complete

danielbachhuber committed:
Encode the query parameter

danielbachhuber committed:
Refactor to merge at access

danielbachhuber commented:
> The person_id ~filter~ dropdown doesn't work:

@mariusandra Good catch, fixed with fdf4a121ded6058d8866790a3a9fde98d88c4126

> As for skipping `checkOrLoadPropertyDefinition`, well, just prepopulate `propertyDefinitionStorage` with the metadata properties and that should be enough, no? One trick you can use is to rename the `propertyDefinitionStorage` reducer to `rawPropertyDefinitionStorage`, and then add a new selector called `propertyDefinitionStorage` that merges `rawPropertyDefinitionStorage` and `eventMetadataPropertyDefinitions` into one object.

Ah, that's quite nice. Thanks for the suggestion! Updated with 070a82df71f6ef06496cec361e9fde255e38a2e9

danielbachhuber commented on frontend/src/lib/taxonomy.tsx:
```diff
@@ -284,6 +284,14 @@ export const CORE_FILTER_DEFINITIONS_BY_GROUP = {
             system: true,
         },
     },
+    event_metadata: {
+        person_id: {
+            label: 'Person ID',
+            description: 'The ID of the person, depending on the person properties mode.',
+            examples: ['16ff262c4301e5-0aa346c03894bc-39667c0e-1aeaa0-16ff262c431767'],
+        },
+        // Other values are assigned below
+    },
```
@mariusandra Do you know how `metadata` is actually used? I was hesitant to use it there because I wasn't confident that I wasn't introducing an undesirable side effect.

mariusandra replied:
Hmm... I'm not too sure, but it looks like it was added in [this PR](https://github.com/PostHog/posthog/pull/20106)

Looking at the code:

![image](https://github.com/user-attachments/assets/727a17e7-45cf-4a59-a141-f3115636dedf)

... it does seem like "distinct_id" was explicitly moved out of the "event_properties" tab because it wasn't a property, but it was more like (event) "metadata". So I think the intent is the same. As for where is it used, I could only find the `CORE_FILTER_DEFINITIONS_BY_GROUP.event_properties.distinct_id = CORE_FILTER_DEFINITIONS_BY_GROUP.metadata.distinct_id` line right now. 

I think it's safe to merge the two... unless we get any immediate typescript errors, probably the worst that can happen is some missing descriptions, which can be fixed later... I also think it's worth merging, as otherwise people will lose time wondering the same question 2 years down the line :D.

danielbachhuber replied:
> So I think the intent is the same. As for where is it used, I could only find the `CORE_FILTER_DEFINITIONS_BY_GROUP.event_properties.distinct_id = CORE_FILTER_DEFINITIONS_BY_GROUP.metadata.distinct_id` line right now.

@mariusandra... I figured out where it's used! It actually didn't take long, only asking the question "where is this used?". I think my mind blocked me from asking that before because I was already part of the way through the implementation when I discovered `TaxonomicFilterGroupType.Metadata`

![CleanShot 2025-03-25 at 05 25 17@2x](https://github.com/user-attachments/assets/281cabab-031f-485b-b03f-0d7bd697c627)

https://github.com/PostHog/posthog/blob/a728bcc1c284f5534794751816add823578d6f8d/frontend/src/scenes/feature-flags/FeatureFlagReleaseConditionsLogic.ts#L291-L295

Given this new knowledge, how do you feel about this PR?

Argument in favor of merging as-is: `TaxonomicFilterGroupType.EventMetadata` is a different use case than `TaxonomicFilterGroupType.Metadata`, and the implementations are a bit different.

Argument against merging as-is: `TaxonomicFilterGroupType.EventMetadata` and `TaxonomicFilterGroupType.Metadata` refer to the same underlying data, so `TaxonomicFilterGroupType.EventMetadata` should be refactored into `TaxonomicFilterGroupType.Metadata`.

mariusandra replied:
Oh wow... that's a bit of a fail for both of us :D. Where is this "metadata" filter used? Just feature flags? I hadn't seen it in the wild, so didn't even think of it :(.

Regarding what to do... let's keep it as is: two lists. The `{ propertyFilterType: PropertyFilterType.Person }` part shows that they're still somewhat different. If we'd merge the two, then every place should/would get the full list (distinct_id + event, timestamp, etc), which I think won't make sense for feature flag matching. Thus we'd need to implement some weird workarounds.

danielbachhuber committed:
Merge branch 'master' into crm/event-metadata

danielbachhuber committed:
Pass group type index through so values aren't nooped

danielbachhuber commented:
> I think I fixed this with [fcfba12](https://github.com/PostHog/posthog/commit/fcfba126d322d171f4ed167b3a02caa14dd74ed7) and [0fda1bc](https://github.com/PostHog/posthog/commit/0fda1bc35814a5290a3005cca1e7f8503b64fdcb) (at least, I don't see the erroneous `property_definitions?limit=50&properties=event&type=event` API request anymore).

Kind of ugly, but I also had to do 1efb7ee4dc. Otherwise, the group filters end up with empty values:

![CleanShot 2025-03-24 at 10 44 04@2x](https://github.com/user-attachments/assets/b9f31b94-b2ef-441a-a875-bb90b16090b0)

![CleanShot 2025-03-24 at 10 44 18@2x](https://github.com/user-attachments/assets/4084a0a5-48d3-4390-b2d3-60664bca6e9e)

mariusandra reviewed (approved):
This has been a big undertaking and thanks for putting up with all my comments and remarks. However I think we made it! 🙌

I still left one comment inline about merging the objects in taxonomy.tsx, however the rest LGTM! So merge whenever you're ready! 👍

danielbachhuber committed:
Merge branch 'master' into crm/event-metadata

danielbachhuber committed:
Move `person_id` into `metadata`

danielbachhuber committed:
Always apply the not null comparison

danielbachhuber committed:
Fix type issue

danielbachhuber committed:
Use static values to prevent racy behavior

danielbachhuber committed:
Revert "Use static values to prevent racy behavior"

This reverts commit bd93bec9b96c1ca276fe2c69515f19e97dc011a9.

danielbachhuber committed:
Update tests for non-deterministic return order
